File: androidApp/src/main/kotlin/it/iterapp/app/common/Formatters.kt
package it.iterapp.app.common

import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import kotlin.math.roundToInt

/**
 * The router's timezone. OTP interprets the `date`/`time` plan arguments in the
 * network's local zone, so departure times must be formatted there, not in the
 * device zone. The served network is Italian; revisit when coverage widens.
 */
private val ROUTER_ZONE: TimeZone = TimeZone.getTimeZone("Europe/Rome")

/** `14:05` in the device locale's 24h clock. */
fun formatClock(epochMs: Long): String =
  SimpleDateFormat("HH:mm", Locale.getDefault()).format(Date(epochMs))

/** `7 min` / `1 h 12 min`. */
fun formatDuration(seconds: Long): String {
  val minutes = (seconds / 60).toInt()
  return when {
    minutes < 60 -> "$minutes min"
    else -> "${minutes / 60} h ${minutes % 60} min"
  }
}

/** `350 m` / `2,4 km` style distance. */
fun formatDistance(meters: Double): String = when {
  meters < 950 -> "${((meters / 10).roundToInt() * 10)} m"
  else -> String.format(Locale.getDefault(), "%.1f km", meters / 1000)
}

/**
 * GTFS station names arrive ALL-CAPS ("MILANO PORTA VENEZIA"); title-case them
 * for display, keeping short railway suffixes ("FS", "FN") intact. Mixed-case
 * names pass through untouched, so the transform is idempotent and safe if the
 * feed is ever fixed upstream. Locale pinned: default-locale lowercase would
 * mangle "MILANO" on Turkish-locale devices.
 */
fun formatStationName(raw: String): String {
  val name = raw.trim()
  if (name != name.uppercase(Locale.ITALIAN)) return name
  return name.split(" ").filter { it.isNotEmpty() }.joinToString(" ") { word ->
    if (word.length <= 2) word
    else word.lowercase(Locale.ITALIAN).replaceFirstChar { it.titlecase(Locale.ITALIAN) }
  }
}

/** Seconds → compact `+3 min` / `+45 s` delay label. */
fun formatDelay(seconds: Double): String {
  val s = seconds.roundToInt()
  return if (s >= 90) "${(s / 60.0).roundToInt()} min" else "$s s"
}

/** Hour and minute of [epochMs] on the router's wall clock. */
fun routerHourMinute(epochMs: Long): Pair<Int, Int> {
  val cal = Calendar.getInstance(ROUTER_ZONE).apply { timeInMillis = epochMs }
  return cal.get(Calendar.HOUR_OF_DAY) to cal.get(Calendar.MINUTE)
}

/** Epoch millis of today at [hour]:[minute] on the router's wall clock. */
fun routerTimeTodayAt(hour: Int, minute: Int): Long =
  Calendar.getInstance(ROUTER_ZONE).apply {
    set(Calendar.HOUR_OF_DAY, hour)
    set(Calendar.MINUTE, minute)
    set(Calendar.SECOND, 0)
    set(Calendar.MILLISECOND, 0)
  }.timeInMillis

/** OTP `YYYY-MM-DD` + `HH:MM` for a plan departure, in the router's zone. */
fun planDateTime(epochMs: Long): Pair<String, String> {
  val dateFmt = SimpleDateFormat("yyyy-MM-dd", Locale.US).apply { timeZone = ROUTER_ZONE }
  val timeFmt = SimpleDateFormat("HH:mm", Locale.US).apply { timeZone = ROUTER_ZONE }
  return dateFmt.format(Date(epochMs)) to timeFmt.format(Date(epochMs))
}

File: iosApp/Sources/Support/Formatters.swift
import Foundation

enum Formatters {
  private static let clockFormatter: DateFormatter = {
    let formatter = DateFormatter()
    formatter.dateFormat = "HH:mm"
    return formatter
  }()

  /// `14:05` from epoch milliseconds (matching the Android 24h clock).
  static func clock(_ epochMs: Int64) -> String {
    clockFormatter.string(from: Date(timeIntervalSince1970: TimeInterval(epochMs) / 1000))
  }

  /// `7 min` / `1 h 12 min`.
  static func duration(_ seconds: Int64) -> String {
    let minutes = Int(seconds / 60)
    if minutes < 60 { return "\(minutes) min" }
    return "\(minutes / 60) h \(minutes % 60) min"
  }

  /// `350 m` / `2.4 km` style distance.
  static func distance(_ meters: Double) -> String {
    if meters < 950 {
      return "\(Int((meters / 10).rounded()) * 10) m"
    }
    return String(format: "%.1f km", locale: .current, meters / 1000)
  }

  /// Seconds → compact `3 min` / `45 s` delay label.
  static func delay(_ seconds: Double) -> String {
    let rounded = Int(seconds.rounded())
    if rounded >= 90 { return "\(Int((Double(rounded) / 60).rounded())) min" }
    return "\(rounded) s"
  }

  /// `312 KB` / `4.2 MB` download progress.
  static func bytes(_ bytes: Int64) -> String {
    if bytes < 1024 * 1024 { return "\(bytes / 1024) KB" }
    return String(format: "%.1f MB", locale: .current, Double(bytes) / (1024 * 1024))
  }
}
